build.sh

```diff
@@ -0,0 +1,27 @@
+#!/usr/bin/env bash
+
+set -ex
+
+ASWF_ORG=${DOCKERHUB_ORG:-aswfstaging}
+CI_COMMON_VERSION=${CI_COMMON_VERSION:-1.0}
+VFXPLATFORM_VERSION=${VFXPLATFORM_VERSION:-2019}
+ASWF_VERSION=${ASWF_VERSION:-${VFXPLATFORM_VERSION}}
+BUILD_DATE=${BUILD_DATE:-dev}
+
+# Also source versions environment variables here as docker image needs some as build arguments
+# Exemple: PYTHON_VERSION
+source scripts/${VFXPLATFORM_VERSION}/versions_base.sh
+
+BUILD_ARGS=" --build-arg BUILD_DATE=${BUILD_DATE} \
+             --build-arg ASWF_ORG=${ASWF_ORG} \
+             --build-arg VFXPLATFORM_VERSION=${VFXPLATFORM_VERSION} \
+             --build-arg CI_COMMON_VERSION=${CI_COMMON_VERSION} \
+             --build-arg PYTHON_VERSION=${PYTHON_VERSION}"
+
+docker build -t ${ASWF_ORG}/ci-base:${ASWF_VERSION} -f ci-base/Dockerfile ${BUILD_ARGS} .
+
+# Most VFX packages are not ready for python3 yet...
+if [[ ${VFXPLATFORM_VERSION} != 2020 ]]; then
+    docker build -t ${ASWF_ORG}/ci-ocio:${ASWF_VERSION} -f ci-ocio/Dockerfile ${BUILD_ARGS} .
+    docker build -t ${ASWF_ORG}/ci-openvdb:${ASWF_VERSION} -f ci-openvdb/Dockerfile ${BUILD_ARGS} .
```

Comment:
I'm wondering if we could perhaps do away with ci-base and move all the dependencies up into ci-ocio/ci-openvdb (and add a ci-openexr)? Having a three-level hierarchy (ci-common -> ci-base -> ci-openexr) seems like unnecessary complexity if it doesn't change very often. Although it's mildly wasteful, at least the large build times (ie Clang) are in the one common image and from what I remember none of the VFX dependencies took very long to compile (is that still the case?).

It feels like that might be simpler to manage if one project wished to experiment with different compiler flags to Boost for example, rather than having to mess with the hierarchy first.

Reply:
I see your point about simplifying the image inheritance. I don't have a strong opinion either way, but here's a summary to discuss:

Build times:
ci-common (system packages, compilers): ~ 2 hours
ci-base (python, boost, tbb, other stuff): ~20m
ci-openvdb (openexr, blosc): ~10m
ci-ocio (openexr, oiio): ~10m

I have added a "patchup.sh" that allows downstream images to install anything they wish, which can be re-installing another version of an upstream to try things, I use it in the 2020 image to update cmake as it officially supports python-3 from 3.12 onwards.

Reply:
One other thing that I wanted to do was to also add Qt to the base images... With Qt-5.12 coming online with 2019 I already see a different build of Qt from every vendor and even different builds of Qt from the same vendor... I'd like to point them to the ci-base as the golden reference for how we need base VFX dependencies built.
I realise that Qt will add both time and size to the base image... Ideas welcome!

Reply:
I suppose Qt creates a good reason for the three-level hierarchy as you don't really want that in the common image or the top-level images.

On the other hand, I don't think any of the existing projects (openvdb, opencolorio, openexr) have a dependency on Qt though? I agree with you that I think it's nice to have an image that has everything in it, but how about:

ci-common -> ci-base
ci-common -> ci-reference
ci-common -> ci-openvdb
ci-common -> ci-ocio
ci-common -> ci-openexr

The ci-base one would contain all the non-vfx third party dependencies as it does at the moment (python, boost, tbb, etc).
The ci-reference one would then contain the full VFX reference platform (python, boost, openexr, openvdb, Qt, etc).
The project ones wouldn't contain Qt.

For each image, we would just duplicate the installation of the python, boost, tbb, etc as it's only ~20m and I think a shallower image hierarchy tree will make things simpler.

Reply:
opencolorio seems to have an optional dependency on Qt, maybe @michdolan can confirm?
I'm happy to go with the flatter hierarchy though, it makes sense for simplicity.

Reply:
The Qt reference in OCIO is in an example python module for an old version of Mari, so nothing needing support in a CI build. 

I'm using ci-base currently in OCIO CI to test a build against "latest" dependency versions. Having an image before any project-specific dependencies are installed is useful for cases like that. That could be a ci-reference image too though. I agree having a golden reference would be a great resource in general.

Reply:
Thanks for the update!

Reply:
@michdolan - it would be great to get a "latest" docker image, we'd really like that too!

This is a bit of a side discussion, but one workflow issue I see with the latest builds as a native part of the CI is that the build isn't necessarily reproducible. If a PR is submitted and everything is green, then an additional commit is added with some feedback addressed, the build may pick up a newer version of a dependency and fail. This could cause confusion to the user thinking it's their commit that has broken the build.

I'm not sure the best practice to resolve this, but I'm thinking of a workflow such as having a weekly build that runs against the latest image and then have PRs run off an earlier "latest" image, one that has (hopefully) already been successful. This would all be managed by tags.

Perhaps a weekly cron could check for any latest versions of any dependency and automatically push a change to trigger a new build generating a new image if needed? Either way, I think it would be great to support latest as a native part of this docker repo...

Reply:
I was also thinking to add "integration tests" for these images so the `ci-openvdb` is used to build the matching vfx platform version of OpenVDB, maybe these could run before release, and would ensure that `aswf/ci-openvdb:latest` should never break opencdb builds...

Reply:
Yeah @danrbailey , that's a good point about running the "latest" build as a cron vs. for PRs. I actually hit that very issue recently where I made another commit to my Azure PR and the OpenEXR master branch build failed. I modified our build so that "latest" pulls the latest tag instead of master to provide some level of stability, but ultimately we'll likely separate it as you suggested.